[DOC][Relay]: Add initial language reference and API docs for Relay.

jroesch opened the pull request:
This PR comprises the work I did on docs for Relay in the initial PR #1672. 

This adds a first pass at the language reference as well as extended documentation for the Python API docs, and some tutorials.

Looking for early feedback on this as we continue to merge the remaining pieces of Relay.

tqchen commented on docs/api/python/relay/op.rst:
```diff
@@ -0,0 +1,3 @@
+tvm.relay.op
+-----------
```
rst requires the underline to havethe same length as the title

tqchen commented on docs/langref/relay/expressions.rst:
```diff
@@ -0,0 +1,161 @@
+==================
+Expressions
+==================
```
underline need to have the same length of the title.
````
First Level
=======
second level
--------------
third level
~~~~~~~
```

MarisaKirisame commented on docs/langref/relay/expressions.rst:
```diff
@@ -0,0 +1,161 @@
+        if (x == 0) {
+            0
+        } else {
+            x * fact(x - 1)
```
missing }

MarisaKirisame commented on docs/langref/relay/expressions.rst:
```diff
@@ -0,0 +1,161 @@
+
+.. code-block:: python
+    if (sum(equal(t, u))) {
+        jreturn x:
```
jreturn -> return

jroesch commented:
@joshpoll @weberlo @MarisaKirisame @slyubomirsky could you guys take passes over the content. I am going to fix/clean up things which changed between last week and the final PR.

tqchen commented:
@junrushao1994 @zhiics @yzhliu @ZihengJiang please review

MarisaKirisame commented on docs/langref/relay/expressions.rst:
```diff
@@ -0,0 +1,162 @@
+
+See :py:class:`~tvm.relay.expr.Call` for its definition and documentation.
+
+If-Then-Else
```
move this before function?

MarisaKirisame commented on docs/langref/relay/intro.rst:
```diff
@@ -0,0 +1,205 @@
+Relay is differentiable programming language with support for
+closures, control-flow, and recursion. It has an advanced
+static type system specifically designed for programs written
+by machine  learning practioniers and researchers.
```
extra whitespace

MarisaKirisame commented on docs/langref/relay/type_system.rst:
```diff
@@ -0,0 +1,187 @@
+
+The shape may be any valid Relay shape as described in the below
+section on shapes.
+
```
extra line

MarisaKirisame commented on docs/langref/relay/type_system.rst:
```diff
@@ -0,0 +1,187 @@
+
+See :py:class:`~tvm.relay.type.IncompleteType` for its definition and documentation.
+
+
```
extra line

MarisaKirisame commented on tutorials/relay/implement_fma_transform.py:
```diff
@@ -0,0 +1,141 @@
+np.testing.assert_allclose(norm_out, xform_out)
+
+# Use the TVM runtime.
+
```
extra line

zhiics commented on docs/langref/relay/intro.rst:
```diff
@@ -0,0 +1,205 @@
+
+*Note: this is not yet implemented.*
+
+Finally we allo a definition be prefixed by metadata, which adds
```
allo? You meant allow?

zhiics commented on docs/langref/relay/type_system.rst:
```diff
@@ -0,0 +1,187 @@
+Type Call
+~~~~~~~~~
+
+Apply a type relation to a set of input arguments, at the present momen the type
```
momen?

zhiics commented on tutorials/relay/implement_fma_transform.py:
```diff
@@ -0,0 +1,141 @@
+    def __init__(self):
+        self.expr_map = {}
+
+    def expr_at(self,id: ir.LocalId) -> ir.Expr:
```
Add space after `self,`

masahi commented on docs/langref/relay/expressions.rst:
```diff
@@ -0,0 +1,162 @@
+is the continutation after executing the sequence of bindings.
+
+I believe this representation will be easier to manipulate then
+the mixed dataflow/control flow comptuation graphs.
```
computation

masahi commented on docs/langref/relay/intro.rst:
```diff
@@ -0,0 +1,205 @@
+beyond the type of identifier.
+
+Globals are written with `@`, locals are written with `%`, and operators are written with
+no sigil like LLVM's intrinsics. The distinction between global and local identifiers
```
sigil?

masahi commented on docs/langref/relay/type_system.rst:
```diff
@@ -0,0 +1,187 @@
+
+See :py:class:`~tvm.relay.type.TypeParam` for its definition and documentation.
+
+Type Constriant
```
Constraint

masahi commented on tutorials/relay/implement_fma_transform.py:
```diff
@@ -0,0 +1,141 @@
+from relay.operators.register import func_ty_to_placeholders, register_op
+from relay.eval import defn_to_pyfunc
+from relay.tyck import check_expr
+
```
I don't see relay.frontend, eval, tyck etc in the current repo. Are they going to be upstreamed soon?

jroesch replied:
This code still needs to be updated from the old branch, but hope is to upstream these pieces in the next few days. 

We have a few parallel efforts going on, some changes to the type checker for supporting VTA, the runtime compiler and evaluator, AD, debugging support, and work towards automatic quantization/8bit support. We will hopefully be releasing most of this in the next couple weeks. 

3 of our team members have been away for the summer so parallelization should pick up soon.

ZihengJiang commented on docs/langref/relay/type_system.rst:
```diff
@@ -0,0 +1,187 @@
+Kind
+~~~~
+The kind of a type parameter, represents a variable shape,
+base type, type, or dimension.
```
what's the difference between `base type` and `type`? Actually, a little confused about this part. Can I understand `Kind` as type's type?

MarisaKirisame replied:
Kind is basically a type's type.
base type is type that has a simple data layout, and thus can act as element of tensor efficiently.

ZihengJiang commented on docs/langref/relay/expressions.rst:
```diff
@@ -0,0 +1,162 @@
+
+We support tuple constructors; the tuple node builds a finite (i.e statically known size) sequence of
+heterogenous data.  These tuples match closely to Python's and enable efficient projection of their m
+embers due to their fixed length.
```
m embers -> members

ZihengJiang commented on docs/langref/relay/expressions.rst:
```diff
@@ -0,0 +1,162 @@
+Function
+~~~~~~~~
+
+A function node represents a function, it contains a seqeuence of
```
sequence

ZihengJiang commented on docs/langref/relay/expressions.rst:
```diff
@@ -0,0 +1,162 @@
+which may reference the bound identifier.
+
+We will first introduce a single binding with no type
+anntoations:
```
annotations

slyubomirsky commented on docs/langref/relay/expressions.rst:
```diff
@@ -0,0 +1,162 @@
+Expressions
+===========
+
+The Relay IR is a pure, expression oriented language, with a dataflow fragment
```
"Expression-oriented" should be hyphenated (compound adjective)

Additionally, I think in the next line it should say "Although Relay's representation is an abstract syntax tree" to clarify what exactly we mean by tree (i.e., not a tree in the sense of an acyclic graph)

slyubomirsky commented on docs/langref/relay/intro.rst:
```diff
@@ -0,0 +1,205 @@
+Introduction
+==================
+
+Relay is differentiable programming language with support for
```
Missing article ("Realy is a differentiable programming language...")

slyubomirsky commented on docs/langref/relay/intro.rst:
```diff
@@ -0,0 +1,205 @@
+used to execute it uncessarily hampers many of the goals
+of a framework.
+
+We believe having a high level, an expressive language designed
```
"a high-level, expressive language..."

slyubomirsky commented on docs/langref/relay/intro.rst:
```diff
@@ -0,0 +1,205 @@
+This may be a useful way to describe early ML models, but is
+limited onceyou want to generalize to dynamic models.
+
+At the same time computation graphs have been over loaded
```
"At the same time, computation graphs have an overloaded purpose as both..."

slyubomirsky commented on docs/langref/relay/intro.rst:
```diff
@@ -0,0 +1,205 @@
+The conflation of the description of a computation, the
+representation used for optimizing it, and the data structure
+used to execute it uncessarily hampers many of the goals
+of a framework.
```
This sentence is phrased rather confusingly. Perhaps the following might be clearer?

"Unnecessarily conflating the description, representation, and execution environment of a computation hampers many of the goals of a framework." (Also appears that this may need to be explained in greater detail.)

slyubomirsky commented on docs/langref/relay/intro.rst:
```diff
@@ -0,0 +1,205 @@
+class functions and closures, enabling the design of differentiable
+arbitrary control flow constructs. Second we contribute a type system for
+Relay which makes the important properties of computations involving tensors
+explicit.
```
Miscellaneous small changes:
* "first-class" in this case should be hyphenated (compound adjective)
* There should be a comma in "the design of differentiable, arbitrary control flow constructs"
* There should be commas after "First" and "Second"
* It might be clearer in the last sentence if "makes the important properties of computations involving tensors
explicit" were "makes explicit the important properties of computations involving tensors"

slyubomirsky commented on docs/langref/relay/intro.rst:
```diff
@@ -0,0 +1,205 @@
+
+We believe Relay will enable researchers and industry to design a
+new class of DL framework frontends, differentiable programming languages,
+deep probablistic programming languages, thus accelerating the pace of
```
Should have "and" before "deep probabilistic programming languages"

slyubomirsky commented on docs/langref/relay/intro.rst:
```diff
@@ -0,0 +1,205 @@
+no sigil like LLVM's intrinsics. The distinction between global and local identifiers
+makes certain kinds of transformation easier. For example inlining a global definition
+requires no analysis, you can write a pass that just directly inlines the defintions.
+It also ensures there are no spooky action at a distance, introducing a new identifier
```
Tense does not match: should be "ensures there are no spooky actions at a distance" or "ensures there is no spooky action at a distance." Additionally, the comma should be a semicolon.

slyubomirsky commented on docs/langref/relay/intro.rst:
```diff
@@ -0,0 +1,205 @@
+return type is omitted we will infer the return type based on the text of the
+program.
+
+Finally we can directly construct type polymorphic definitions by writing down
```
should be a hyphen in "type-polymorphic"

slyubomirsky commented on docs/langref/relay/intro.rst:
```diff
@@ -0,0 +1,205 @@
+to parameters in the machine learning.
+
+When the type information is omitted we will attempt to infer a most general type
+for the users. This property is known as generalization, for a definition without
```
I think it would be clearer if the comma after "generalization" were a colon (can misinterpret the "for" in the current phrasing)

slyubomirsky commented on docs/langref/relay/type_system.rst:
```diff
@@ -0,0 +1,187 @@
+
+Static types are useful when performing compiler optimization because they
+communicate properties about the data we manipulate, such as runtime shape,
+data layout, storage without needing to run the program.
```
Missing conjunction before "storage"

slyubomirsky commented on docs/langref/relay/type_system.rst:
```diff
@@ -0,0 +1,187 @@
+community, and can be viewed as a method for generalizing shape inference to
+run over arbitrary user programs containing control flow and recursion.
+
+Static types are useful when performing compiler optimization because they
```
"compiler optimizations" should be plural

slyubomirsky commented on docs/langref/relay/type_system.rst:
```diff
@@ -0,0 +1,187 @@
+===========
+
+We have briefly introduced types while detailing the the expression language
+of Relay, but have fully laid out the type system.
```
Sounds like it should say "but have not yet fully laid out the type system"

slyubomirsky commented on docs/langref/relay/intro.rst:
```diff
@@ -0,0 +1,205 @@
+~~~~~~~~
+
+Now that we have presented both global functions, and operators we have
+everthing in hand to describe a complete Relay program. A Relay program consists of a
```
typo: "everything"

slyubomirsky commented on docs/langref/relay/intro.rst:
```diff
@@ -0,0 +1,205 @@
+Operators
+=========
+
+A primitive operation that is not defined in the Relay language but provided
```
The first sentence here is a fragment. Perhaps it should read, "An operator is a primitive operation that is not defined in the Relay language but is provided externally"?

slyubomirsky commented on docs/langref/relay/intro.rst:
```diff
@@ -0,0 +1,205 @@
+of the operator, its type and various attributes required by Relay subsystems.
+
+The input methods for Relay programs do not provide a way to describe operators in
+Relay, they must be explicitly registered in the global environment via Python or C++.
```
the comma after "Relay" should be a semicolon

slyubomirsky commented on docs/langref/relay/intro.rst:
```diff
@@ -0,0 +1,205 @@
+A primitive operation that is not defined in the Relay language but provided
+externally. Currently we back these operators registrations with the operators
+exposed by TVM's TOPI. An operator requires a user to provide an implementation
+of the operator, its type and various attributes required by Relay subsystems.
```
Missing an Oxford comma (they're used everywhere else consistently)

slyubomirsky commented on docs/langref/relay/intro.rst:
```diff
@@ -0,0 +1,205 @@
+=========
+
+A primitive operation that is not defined in the Relay language but provided
+externally. Currently we back these operators registrations with the operators
```
missing apostrophe, namely "operators' registrations"

slyubomirsky commented on docs/langref/relay/type_system.rst:
```diff
@@ -0,0 +1,187 @@
+communicate properties about the data we manipulate, such as runtime shape,
+data layout, storage without needing to run the program.
+
+Most current IRs use "shape inference" to recover Tensor dimensions from the user
```
"tensor" is unnecessarily capitalized

slyubomirsky commented on docs/langref/relay/type_system.rst:
```diff
@@ -0,0 +1,187 @@
+higher level inferences.
+
+We view this limited shape inference as a simpler form of type
+inference. Instead of relying on an ad-hoc procedure for recovering type
```
"ad hoc" is generally not hyphenated

slyubomirsky commented on docs/langref/relay/type_system.rst:
```diff
@@ -0,0 +1,187 @@
+Types
+=====
+
+Relay's type system has a "language of types" which allow us to write down the type of
```
"allows us" rather than "allow us"

slyubomirsky commented on docs/langref/relay/type_system.rst:
```diff
@@ -0,0 +1,187 @@
+=====
+
+Relay's type system has a "language of types" which allow us to write down the type of
+a Relay program. Below we detail the langauge of types and how we assign them to Relay
```
"them" is ambiguous in this sentence, would be clearer to explicitly write "how we assign types"

slyubomirsky commented on docs/langref/relay/type_system.rst:
```diff
@@ -0,0 +1,187 @@
+This is the type assigned to functions in Relay. They consist of
+a list of type parameters which enable the definition of generic
+functions, a set of type constraints which we omit for the time
+being, a sequence of argument types, and a return type.
```
"which we omit for the time being" is a parenthetical remark, so it should be offset from the surrounding context with parentheses or dashes

slyubomirsky commented on docs/langref/relay/type_system.rst:
```diff
@@ -0,0 +1,187 @@
+A type relation is the most exotic type system feature in Relay. It allows
+users to extend type and shape checking/inference with new rules. We use
+type relations to type operators with "hard" types such as broadcasting
+operators, or special ones like :code:`flatten`.
```
the comma appears to be unnecessary

slyubomirsky commented on docs/langref/relay/type_system.rst:
```diff
@@ -0,0 +1,187 @@
+type relations to type operators with "hard" types such as broadcasting
+operators, or special ones like :code:`flatten`.
+
+A type relation :code:`R` is a n-ary input, single output relation over
```
there should be a hyphen in "single-output"

slyubomirsky commented on docs/langref/relay/type_system.rst:
```diff
@@ -0,0 +1,187 @@
+
+Apply a type relation to a set of input arguments, at the present momen the type
+call node represents the application of a :py:class:`~tvm.relay.type.TypeRelation`
+to a set of input arguments. The result of type application is the output variable
```
Missing article: "The result of a type application..."

slyubomirsky commented on docs/langref/relay/expressions.rst:
```diff
@@ -0,0 +1,162 @@
+Dataflow Expressions
+====================
+
+First we will cover the set of nodes which do not involve control flow,
```
the comma after "control flow" should be a semicolon (or the clauses should be made separate sentences entirely)

slyubomirsky commented on docs/langref/relay/expressions.rst:
```diff
@@ -0,0 +1,162 @@
+a constant tensor value (values are either Tensors, Products, or Closures in Relay).
+The constants are represented as :py:class:`~tvm.NDArray`, allowing us to utilize
+TVM operators for constant evaluation.
+
```
Unnecessary capitalization in the above paragraph ("Tensor," "Products," "Closures")

joshpoll commented on docs/api/python/relay/index.rst:
```diff
@@ -0,0 +1,20 @@
+This document contains the Python API to the Relay frontend, optimizer, and
+compiler toolchain.
+
+Relay is the second generation high level intermediate representation for the TVM 
```
there should also be a comma:

"Relay is the second-generation, high-level intermediate representation for the TVM compiler stack."

ZihengJiang commented on tutorials/relay/implement_fma_transform.py:
```diff
@@ -0,0 +1,141 @@
+
+class FuseTwo(ExprAtVisitor):
+    """Rewrite b(a(x, y), z) into ab(x, y, z). """
+    def __init__(self, a: OperatorId, b: OperatorId, a_and_b: OperatorId) -> None:
```
Do we have doc and explanation for `OperatorId`?

ZihengJiang commented on tutorials/relay/implement_fma_transform.py:
```diff
@@ -0,0 +1,141 @@
+        func = call.fn
+        if func == self.b:
+            assert len(call.args) == 2 # An assumption of this fusion code.
+            arg0 = self.expr_at(call.args[0])
```
what is `expr_at`? why can't we do `arg0 = call.args[0]` directly?

ZihengJiang commented on tutorials/relay/implement_fma_transform.py:
```diff
@@ -0,0 +1,141 @@
+    shape = mk.TypeParam("s", ir.Kind.Shape)
+    bt = mk.TypeParam("bt", ir.Kind.BaseType)
+    in_out_type = mk.TensorType(bt, shape)
+    fma_type = mk.TypeQuantifier(bt, mk.TypeQuantifier(shape, mk.TypeArrow([in_out_type, in_out_type, in_out_type], in_out_type)))
```
This expression looks tedious and hard to understand. Can you explain a bit more?

tqchen commented on docs/langref/relay/expressions.rst:
```diff
@@ -0,0 +1,162 @@
+===========
```
Let us move directly into docs/langref/relay_expr.rst

The future trend will just make everything as a one project and having less subfolder will help to do so

tqchen commented:
@jroesch given the current status, let us consider remove the tutorial part and send it as a followup PR and aim to merge the document part quickly. GIven that there is not a lot of language reference, let us just put all docs under docs/langref/relay_intro, docs/langref/relay_expr, to reduce amount of nests

yzhliu commented:
@jroesch Any update on the changes?

jroesch commented:
@yzhliu Hey sorry, I was on vacation for the past two weeks will put some cycles on this today.

tqchen commented:
Some comments:

- Let us move langref/relay/*.rst to langref/relay_*.rst, as there won't be a lot language reference and it reduces one level of indirection.
- When writing IR code, maybe we want to think about consistency with the text format.
- Incorporate some of our discussions of graph mode vs let binding with text format will really help explain the concept

slyubomirsky commented on docs/langref/relay/intro.rst:
```diff
@@ -0,0 +1,182 @@
+sigil name the corresponding operator. The distinction between global and local identifiers
+makes certain kinds of transformation easier. For example, inlining a global definition
+requires no analysis, as you can write a pass that directly inlines the definitions.
+Ensuring there is no spooky action at a distance; introducing a new identifier return
```
Was this sentence clobbered during a merge or rebase? It no longer makes any sense and I didn't know what to change it to

jroesch commented:
Hey everyone @tqchen @zhiics @slyubomirsky @junrushao1994 @zhiics @yzhliu @ZihengJiang sorry this took so long to update, things got busy working on the code. Split out just the API pieces and updated the pull request, if people could just give a quick review it would be great to merge soon. My final todo is fixing up cross-references to work more effectively.

zhiics commented:
@jroesch LGTM.
